internal/controllers/core/kubernetesapply/disco.go

```diff
@@ -0,0 +1,181 @@
+package kubernetesapply
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/tilt-dev/tilt/internal/controllers/apicmp"
+	"github.com/tilt-dev/tilt/internal/k8s"
+	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
+)
+
+// Each KubernetesApply object owns a KubernetesDiscovery object of the same name.
+//
+// After we reconcile a KubernetesApply, update the KubernetesDiscovery objects it owns.
+//
+// If the Apply has been deleted, any corresponding Disco objects should be deleted.
+func (r *Reconciler) manageOwnedKubernetesDiscovery(ctx context.Context, nn types.NamespacedName, ka *v1alpha1.KubernetesApply) error {
+	if !r.enableDiscoForTesting {
+		// TODO(nick): remove this.
+		return nil
+	}
+
+	if ka != nil && (ka.Status.Error != "" || ka.Status.ResultYAML == "") {
+		// If the KubernetesApply is in an error state or hasn't deployed anything, don't
+		// reconcile the discovery object. This prevents the reconcilers
+		// from tearing down all the discovery infra on a transient deploy error.
+		return nil
+	}
+
+	var existingKD v1alpha1.KubernetesDiscovery
+	err := r.ctrlClient.Get(ctx, nn, &existingKD)
+	isNotFound := apierrors.IsNotFound(err)
+	if err != nil && !isNotFound {
+		return fmt.Errorf("failed to fetch managed KubernetesDiscovery objects for KubernetesApply %s: %v",
+			ka.Name, err)
+	}
+
+	kd, err := r.toDesiredKubernetesDiscovery(ka)
+	if err != nil {
+		return fmt.Errorf("generating kubernetesdiscovery: %v", err)
+	}
+
+	if isNotFound {
+		if kd == nil {
+			return nil // Nothing to do.
+		}
+
+		err := r.ctrlClient.Create(ctx, kd)
+		if err != nil {
+			return fmt.Errorf("creating kubernetesdiscovery: %v", err)
+		}
+		return nil
+	}
+
+	if kd == nil {
+		err := r.ctrlClient.Delete(ctx, &existingKD)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf("deleting kubernetesdiscovery: %v", err)
+		}
+		return nil
+	}
+
+	if !apicmp.DeepEqual(existingKD.Spec, kd.Spec) {
```

Comment:
ya, Milas and I had a good discussion on this. Basically:

- Comparing ObjectMeta won't work right, since there are fields that are auto-incrementing (e.g., ResourceVersion)
- Agreed that we need a utility function for fields that should cause an update
- We should probably migrate towards controllerutil.CreateOrUpdate and CreateOrPatch
https://github.com/kubernetes-sigs/controller-runtime/blob/v0.9.2/pkg/controller/controllerutil/controllerutil.go#L195
rather than write this code ourselves, though this requires some changes to how we structure the code, and might have gotchas

Reply:
filed https://app.clubhouse.io/windmill/story/12387/use-createorupdate-createorpatch-to-do-updates